### ISSUE TITLE
Unbalanced entity category & too small embedding parameters

### ISSUE THREAD

**Issue by pnnngchg** [facebookresearch/PyTorch-BigGraph#49]
Hello,
First of all, thank you for this amazing project.

I have employed biggraph to train embedding vectors, but the result was not satisfied. So I come to ask for you advice.

The graph I used has 2 entity categories, say, items and users. But the amount of them is quite different.  There are 700,000 items and 10,000,000 users.  The graph has 70,000,000 edges.  Only 1 relationship in there, a user clicked a item. Vector dimension is set as 128,  I trained 30 epoch.

The final embedding vectors also confused me a lot, which have too small parameters, like this:
\46XC28261954309     0.000960 -0.002464 0.001338 -0.000077 0.001412 -0.000633 0.002548 -0.000783 0.001233......

So, I want to ask two questions:
1. Dose the unbalanced entity category influence the result of BigGraph's ? Or to say, can this kind of graph  be properly trained to get good embedding vectors ?
2. Why the vectors I got have too small parameters? I am sure I did not change BigGraph's model configuration. And, I also want to know, dose this small vector parameters influence the quality of embedding vectors?

Looking for your reply,
Thanks!

**Comment by lerks** [comment]
What issue exactly do you see in small embedding components?

I must admit that I never really looked at the magnitudes of the vectors that PBG learned, but I don't see an a priori problem with them being small. For example, if you use the cosine distance as a comparator then the score between a pair of entities only depends on the angle between them, totally disregarding their norm (this stops being true if one uses an operator). Also, if you have the `max_norm` option set then this may artificially keep your vectors small.

In order to know whether you are getting good embeddings for your task you should proceed as usual: withhold a validation/testing set from training, come up with a good metric for evaluation (the `torchbiggraph_eval` command offers [some](https://torchbiggraph.readthedocs.io/en/latest/evaluation.html), but for special cases you may want different ones) and check if the embeddings you're training score high enough on it.

As for your concerns about the imbalance between the two sides, I don't see an issue with that: when constructing negatives PBG will sample entities from the correct entity type, so it won't be biased/diluted by other types being larger or smaller. In your case, since you have just one relation type, I really don't think this should cause you any problems. One other thing that I'm worried about is that you only have 7 edges on average per user, which isn't that many and may lead to poor results. You could consider looking at the distribution of degree by user and see if it may make sense to prune users with low degree hoping to get more signal for those with high degree.

**Comment by lerks** [comment]
No follow-up for a while, closing this. Please reopen if needed.
